recipes/lm/sft/dataset.py

```diff
@@ -0,0 +1,374 @@
+from fairseq2.data.tokenizers import Tokenizer
+from fairseq2.data.tokenizers.hg import HuggingFaceTokenEncoder
+from fairseq2.datasets import DataPipelineReader, SequenceBatch, SyncMode
+from fairseq2.error import NotSupportedError
+from fairseq2.gang import Gangs
+
+from .utils import (
+    Batching,
+    DatasetLoadError,
+    LengthBatching,
+    StaticBatching,
+    load_files_and_weights,
+)
+
+# TODO: FIX, INFER
+npc = 10
+
+
+LM_SFT_DATASET: Final = "lm_sft"
+
+
+@dataclass(kw_only=True)
+class DataReadOptions:
+    batching: Batching = field(default_factory=lambda: StaticBatching(1))
+    """The batching strategy for returned examples."""
+
+    example_shuffle_window: int = 1
+    """
+    The size of the sliding window for shuffling examples. If ``1``, no
+    shuffling is performed; if ``0``, true shuffling is performed by loading the
+    entire dataset.
+    """
+
+    batch_shuffle_window: int = 1
+    """
+    The size of the sliding window for shuffling batches. If ``1``, no
+    shuffling is performed; if ``0``, true shuffling is performed by loading the
+    entire dataset.
+    """
+
+    drop_remainder: bool = False
+    """
+    If ``True``, drops the last set of batches if they have in total fewer
+    examples than requested.
+    """
+
+    sync_batches: bool = True
+    """
+    If ``True``, ensures that each process in the gang reads the same number of
+    batches. Typically used when the amount of data to be read can vary per
+    process (e.g. due to unbalanced sharding or non-static batching) and it is
+    critical for each process to iterate over the same number of batches (e.g.
+    during training).
+    """
+
+    sync_mode: SyncMode = SyncMode.UNTIL_FIRST
+    """
+    The data synchronization mode among processes in the gang. Only effective if
+    :attr:`sync_batches` is ``True``.
+    """
+
+    max_num_batches: int | None = None
+    """The maximum number of batches to return."""
+
+    num_accumulate: int = 1
+    """
+    The number of batches to accumulate in each iteration. Typically used with
+    gradient accumulation during training.
+    """
+
+    prefetch: int = 1
+    """The number of batches to prefetch in background."""
+
+    npc: int = 10
+    """The reference number of parallel calls that data reader can do."""
+
+    seed: int = 2
+    """The seed to initialize the random number generators used internally."""
+
+    extras: MutableMapping[str, object] = field(default_factory=dict)
+    """The reader-specific extra options."""
+
+
+@dataclass(kw_only=True)
+class InstructionReadOptions(DataReadOptions):
+    sample: bool = False
+    """
+    If ``True``, instruction sources (e.g. JSONL files) will be sampled in
+    proportion to their weights.
+    """
+
+    source_encode_mode: str = "prompt"
+    """The tokenizer mode to encode the source text."""
+
+    target_encode_mode: str = "prompt_response"
+    """The tokenizer mode to encode the target text."""
+
+    chat_mode: bool = False
```

Comment:
maybe we don't need this class anymore and it's all just in [Class DataReadOptions](https://github.com/facebookresearch/fairseq2/pull/1271/files#diff-fc9871ae1e62201b728f966210cf78f11734a2d773bc51133a3bc955e20a6f98R48)?